### DISCUSSION TITLE
How to add dimensionless constraints - Π theorem

### DISCUSSION THREAD

**Discussion by Alexia369** [MilesCranmer/PySR#635] [Q&A]
Hello, how can I use the custom loss function to
(1) obtain the dimension exponential vector of the variable
(2) If I can obtain it, I can add the dimensionless constraint -Π theorem to get the desired dimensionless parameter
I am not very good at writing Julia code, and I always get an error when using Unitful.Dimensions()
I would like to ask you how to do it, thank you
<img width="570" alt="屏幕截图_20240602_114149" src="https://github.com/MilesCranmer/PySR/assets/132361044/d9fcb77e-259c-4185-83c7-9dcac22ed5bf">

**Comment by Alexia369** [comment]
`dimensionless_constants_only= True `
I wonder if this line of code has converged to or found a trend for dimensionless parameters，i guess

**Reply by MilesCranmer** [reply]
Yep that’s the parameter. Does that answer the problem or did you still want to use a custom loss?

**Reply by MilesCranmer** [reply]
This is where the dimensionality check occurs in the default loss function: https://github.com/MilesCranmer/SymbolicRegression.jl/blob/71447ee7bee18fc836a7d7bb4c8ce5d249bc0e7c/src/LossFunctions.jl#L71

And this is where dimensional constraint violations are calculated: https://github.com/MilesCranmer/SymbolicRegression.jl/blob/71447ee7bee18fc836a7d7bb4c8ce5d249bc0e7c/src/DimensionalAnalysis.jl#L195

Note that it uses DynamicQuantities; not Unitful

**Reply by Alexia369** [reply]
I referred to the DynamicQuantities.jl document you provided, which mentioned that umass ulength can obtain the dimensional index, but it still reports an error when actually used. Why is that?
```
julia> umass(x)
1//1
julia> ulength(x)
1//2
```

`dim = [DynamicQuantities.ulength(s[1]), DynamicQuantities.utime(s[1])]`
Error reason: MethodError: no method matching ulength(::Float32)
Can you please guide me on how to write this Julia code? This is very important to me. Thank you very much

**Reply by Alexia369** [reply]
Yes，I still want to use a custom loss

**Reply by MilesCranmer** [reply]
`X` does not have units; they need to be multiplied in from `X_units`. And `DynamicQuantities` is not imported automatically in the loss function, you would need to do that separately.
